### ISSUE TITLE
Ray Tune: `BayesOptSearch` config hashing unnecessarily skips suggested trials.

### ISSUE THREAD

**Issue by dataframing** [ray-project/ray#50315]
### What happened + What you expected to happen

Hi all,

I'm using Ray Tune's integration with the `BayesianOptimization` package via `BayesOptSearch` ([link to docs](https://docs.ray.io/en/latest/tune/examples/bayesopt_example.html)). To start, Ray Tune doesn't support the latest version of `BayesianOptimization` ([v2.0.3](https://github.com/bayesian-optimization/BayesianOptimization/releases/tag/v2.0.3)), so we have to downgrade to v1.5.1. My main issue, however, is how `BayesOptSearch` determines whether two sampled trials are duplicates. Specifically, after identifying a suggested `config`, we hash the dictionary so that we can check if we've already computed that value:

https://github.com/ray-project/ray/blob/6037e33c1ccd61dfd6335f160bb4423008ad840f/python/ray/tune/search/bayesopt/bayesopt_search.py#L275-L281

The problem here is that `_dict_hash` encodes floating point parameters to a certain `precision`, which `BayesOptSearch` sets to 5 by default:

https://github.com/ray-project/ray/blob/6037e33c1ccd61dfd6335f160bb4423008ad840f/python/ray/tune/search/bayesopt/bayesopt_search.py#L163-L164

https://github.com/ray-project/ray/blob/6037e33c1ccd61dfd6335f160bb4423008ad840f/python/ray/tune/search/bayesopt/bayesopt_search.py#L35

This seems like a very poor default, especially when dealing with floating point values. For example, if a user is only tuning a value like a neural network's learning rate, this precision would encode `1.46532e-05` and `1.00001e-05` as the same string: `"0.00001"`. As a user, that's pretty confusing. Instead, it seems prudent to either (1) increase the default precision to something more reasonable, like 8 or 16, or (2) allow the user to provide a precision (and call out the implications in the docs). It's fairly easy to manually set `repeat_float_precision = 16` after initialization, but it's just a bit annoying.

Let me know if you have any questions. Thank you!

### Versions / Dependencies

`ray[tune]==2.9.0`

### Reproduction script

(Sorry it's a bit longer than 50 lines; these methods are defined under private paths. I've added my own print statements.)

```python
import copy
import json
from typing import Dict

def flatten_dict(
    dt: Dict,
    delimiter: str = "/",
    prevent_delimiter: bool = False,
    flatten_list: bool = False,
):
    """Flatten dict.

    Output and input are of the same dict type.
    Input dict remains the same after the operation.
    """

    def _raise_delimiter_exception():
        raise ValueError(
            f"Found delimiter `{delimiter}` in key when trying to flatten "
            f"array. Please avoid using the delimiter in your specification."
        )

    dt = copy.copy(dt)
    if prevent_delimiter and any(delimiter in key for key in dt):
        # Raise if delimiter is any of the keys
        _raise_delimiter_exception()

    while_check = (dict, list) if flatten_list else dict

    while any(isinstance(v, while_check) for v in dt.values()):
        remove = []
        add = {}
        for key, value in dt.items():
            if isinstance(value, dict):
                for subkey, v in value.items():
                    if prevent_delimiter and delimiter in subkey:
                        # Raise if delimiter is in any of the subkeys
                        _raise_delimiter_exception()

                    add[delimiter.join([key, str(subkey)])] = v
                remove.append(key)
            elif flatten_list and isinstance(value, list):
                for i, v in enumerate(value):
                    if prevent_delimiter and delimiter in subkey:
                        # Raise if delimiter is in any of the subkeys
                        _raise_delimiter_exception()

                    add[delimiter.join([key, str(i)])] = v
                remove.append(key)

        dt.update(add)
        for k in remove:
            del dt[k]
    return dt


def _dict_hash(config, precision):
    print(f"{config=}; {precision=}")
    flatconfig = flatten_dict(config)
    print(f"{flatconfig=}")
    for param, value in flatconfig.items():
        if isinstance(value, float):
            flatconfig[param] = "{:.{digits}f}".format(value, digits=precision)

    print(f"{flatconfig=}")
    hashed = json.dumps(flatconfig, sort_keys=True, default=str)
    print(f"{hashed=}")
    return hashed


config_1 = {"learning_rate": 1.00001e-05}
config_2 = {"learning_rate": 1.46532e-05}
repeat_float_precision = 5

print(f"=== Config Hash 1 ===")
config_hash_1 = _dict_hash(config_1, repeat_float_precision)
print(f"=== Config Hash 2 ===")
config_hash_2 = _dict_hash(config_2, repeat_float_precision)
assert config_hash_1 == config_hash_2
```

### Issue Severity

Low: It annoys or frustrates me.
